A80-pid.md

```diff
@@ -0,0 +1,312 @@
+    // Ignore empty load
+    return -1
+  }
+  errorRate = load.Eps / load.RpsFractional
+  useErrPenalty = errorRate > conf.ErrorUtilizationThreshold
+  if useErrPenalty {
+    utilization += errorRate * conf.ErrorUtilizationPenalty
+  }
+
+  // Ensure at least WeightUpdatePeriod has passed since the last update.
+  // Prevents corruption of PID controller's internal state, which could happen in the following cases:
+	// * If 2 updates are very close to each other in time, samplingInterval ~= 0 and signal ~= infinity.
+	// * If multiple updates happened during a single WeightUpdatePeriod, the actual weights are not applied,
+	// but PID controller keep growing the weight and it may easily pass the balancing point.
+	if time.Since(lastApplied) < conf.WeightUpdatePeriod {
+		return -1
+	}
+
+  // use value calculated in the onEDFSchedulerUpdate method
+	meanUtilization = data.meanUtilization
+
+  // call PID controlelr to get the value of the control signal.
+	controlSignal = data.pidController.update({
+		referenceSignal:  meanUtilization,
+		actualSignal:     utilization,
+		samplingInterval: time.Since(lastApplied),
+	})
+
+	// Normalize the signal.
+	// If meanUtilization ~= 0 the signal will be ~= 0 as well, and convergence will becoma painfully slow.
+	// If, meanUtilization >> 1 the signal may become very high, which could lead to oscillations.
+	if meanUtilization > 0 {
+		controlSignal *= 1 / meanUtilization
+	}
+
+  lastAppliedWeight = data.lastAppliedWeightPerSubchannel[subchannelID]
+
+  // Use controlSignal to adjust the weight.
+  // First caclulate multiplier that will be used to determine how much weight should be changed.
+  // The higher is the absolute value of the controlSignal the more we need to adjust the weight.
+	if controlSignal >= 0 {
+    // in this case mult should belong to [1,inf) interval, so we will be increasing the weight.
+		mult = 1.0 + controlSignal
+	} else {
+    // in this case mult should belong to (0, 1) interval, so we will be decreasing the weight.
+		mult = -1.0 / (controlSignal - 1.0)
+	}
+	weight = lastAppliedWeight * mult
+
+  // Clamp weight
+	if weight > conf.MaxWeight {
+		weight = conf.MaxWeight
+	}
+	if weight < conf.MinWeight {
+		weight = conf.MinWeight
+	}
+
+  // Save resulting utilization and weight.
+	data.utilizationPerSubchannel[subchannelId] = utilization
+  data.lastAppliedWeightPerSubchannel[subchannelID] = weight
+
+  return weight
+}
+
+func onEDFSchedulerUpdate(data callbackData) {
+  // Calculate mean utilization across all subchannels
+  totalUtilization = 0
+  count = len(data.utilizationPerSubchannel)
+  for _, utilization in data.utilizationPerSubchannel {
+    totalUtilization += utilization
+  }
+  data.meanUtilization = totalUtilization / count
+}
+```
+
+### Dealing with Oscillations
+
+One of the main challenges with the pid balancer is the potential for oscillations. Several factors influence this likelihood:
+
+1. **Propagation Delay of Load Reports:**
```

Comment:
Aside: this is where I think we should consider extending our load metric to support a client's local view of load. This way if a server becomes unresponsive for a period, we can use our previous marker of load together with the clients view. This should help fill in delays in load reports, for example. In general, this sort of approach is much more holistic and takes into account the fact that each client may have a unique network path to the respective server.

Reply:
That's the purpose of wrr blackout period, but I agree that we could make it better with client-side metrics. This probably should go into wrr itself as it is not specific to pid version.